### ISSUE TITLE
Encrypt User Password at Rest

### ISSUE THREAD

**Issue by wesker-albert** [nicotine-plus/nicotine-plus#2834]
### Describe the solution you'd like

I'm unsure if this is the case for all OS, but I've noticed that on Linux, user passwords are stored in plaintext in the config file. This makes me a bit uncomfortable, and puts the user at unnecessary risk.

It'd be nice if the password were salted and hashed, using [`passlib`](https://passlib.readthedocs.io/en/stable/) or similar.

### Additional context

N/A

**Comment by wesker-albert** [comment]
`hashlib` would likely be the pure python solution.

**Comment by mathiascode** [comment]
We can't change the Soulseek protocol or server, so hashing wouldn't be possible. The best we could do is store the password in the system keyring.

**Comment by wesker-albert** [comment]
> We can't change the Soulseek protocol or server, so hashing wouldn't be possible. The best we could do is store the password in the system keyring.

I suppose what I meant was hash it in local storage, then decrypt the hash before authorizing with the server.

To your point though, keyring seems like it could be a more elegant solution.

**Comment by hboetes** [comment]
Even if you store it securely on the system, the password will be transmitted unencrypted over the network as soon as you log in. If you don't trust your local admin, you shouldn't use the system.

**Comment by cheesetosti** [comment]
> Even if you store it securely on the system, the password will be transmitted unencrypted over the network as soon as you log in. If you don't trust your local admin, you shouldn't use the system.

Personally I don't store my money inside my house on the table in broad daylight. But in this case it might be a more cosmetic/principal issue  to not store login credentials in a config file

**Comment by mathiascode** [comment]
The config file has file permissions set to only be readable by the current user. All methods have drawbacks in any case, you can't guarantee that the keyring will be available on less popular operating systems, containers, servers etc.

**Comment by yushisora** [comment]
Is it possible to add a config option like `passw-cmd` to fetch password from a command?

**Comment by slook** [comment]
Yes, you could write a plugin that uses the [loaded_notification()](https://github.com/nicotine-plus/nicotine-plus/blob/7a6efa88e7ea8d02ece926cd4d939621b3fcb032/pynicotine/pluginsystem.py#L56-L58) to fetch a string value from some external command and then assign this into `config.sections.["server"]["passw"]` at startup, then use [shutdown_notification()](https://github.com/nicotine-plus/nicotine-plus/blob/7a6efa88e7ea8d02ece926cd4d939621b3fcb032/pynicotine/pluginsystem.py#L68-L70) to clear it.

This would work because the plugin system loads before automatically connecting to the server at startup, and the Setup Assistant dialog won't launch provided that the `passw` setting isn't empty when a connection attempt is made.
